blog/2020-05-29-mutagen/2020-05-29-mutagen.md

```diff
@@ -0,0 +1,232 @@
+`m` means that the test passed. Therefore `UPDATE_NOTHING` is relevant because no test spotted it
+and similarly `__DELITEM__NOTHING` only made one test fail.
+
+### Step 2: Focus on interesting mutants, make them more robust.
+
+When a mutant is detected by one or two tests we can then find out why they fail and try to make the
+mutant more subtle so the bugs that it introduces are no longer caught. For example in the case of
+`__DELITEM__NOTHING`, we understand that the only failing test function expects an exception when
+the item is not in the dict. Therefore if we replace the `__delitem__` function with a function that
+does nothing if the key exists and else raises a KeyError it passes all tests, even though it's far
+from its expected functioning:
+
+```python
+@mg.mutant_of("MutableBidict.__delitem__", "__DELITEM__NOTHING")
+def __delitem__nothing(self, key):
+    if self.__getitem__(key):
+        return
+    raise KeyError
+```
+
+![](assets/bidict_final_mutations.png)
+
+We now have two undetected mutants!  This means we need more tests.
+
+### Step 3: Add new tests to fill the holes.
+
+The final step is to create new tests that catch the mutants. This is not as easy as it might seem,
+because simply writing a very specific test that catches one particular mutant is unlikely to
+make our test suite more robust. Instead, we need to fundamentally understand what holes in the test
+suite these mutants have unveiled and try to fill them in the most general way possible. Ideally,
+the tests that are added should fit with the other ones and be able to detect other mutants.
+
+In our research group, we are partial to *property-based testing* (PBT), which works especially well
+with mutation testing (although unit tests are fine too). The idea behind PBT is pretty simple:
+rather than test particular cases, tests verify *properties* or invariants. For example, one
+property that might be useful for a dictionary is "a key that was just deleted should not be in the
+dictionary." In Python this is possible with the help of the Hypothesis library, which introduces
+generators for different data types and runs each test function many times with generated inputs to
+validate the property.
+
+For `bidict` we added two correctness properties, and voila, now all the mutants are caught:
+
+```python
+# Slightly simplified version of both properties, for readability reasons
+@given(st.MUTABLE_BIDICTS, st.PAIRS)
+def test_update_correct(mb, pair):
+    key, value = pair
+
+    mb.update([pair])
+    assert mb[key] == value
+
+@given(st.MUTABLE_BIDICTS)
+def test_del_correct(mb):
+    rand_key = choice(list(mb.keys()))
+
+    del mb[rand_key]
+    assert not rand_key in mb
+```
+
+![](assets/bidict_with_new_props.png)
+
+We can then run other mutants to check whether the properties we've added actually catch some of
+them:
+
+![](assets/bidict_all_mutants.png)
+
```

Comment:
I was slightly confused by _belong to the test suite_ where I expected maybe _belong in the test suite_.